src/syntax/sugar.rkt

```diff
@@ -135,97 +140,218 @@
+    (match (impl-info impl 'fpcore)
+      [(list '! _ ... body) body]
+      [body body]))
+  (define subst (pattern-match pattern (cons op args)))
+  (pattern-substitute (cons impl vars) subst))
 
-;; Translates from FPCore to an LImpl
+;; Translates from FPCore to an LImpl.
 (define (fpcore->prog prog ctx)
-  (define-values (expr* _)
-    (let loop ([expr (expand-expr prog)]
-               [ctx ctx])
-      (match expr
-        [`(FPCore ,name (,vars ...) ,props ... ,body)
-         (define-values (body* repr*) (loop body ctx))
-         (values `(FPCore ,name ,vars ,@props ,body*) repr*)]
-        [`(FPCore (,vars ...) ,props ... ,body)
-         (define-values (body* repr*) (loop body ctx))
-         (values `(FPCore ,vars ,@props ,body*) repr*)]
-        [`(if ,cond ,ift ,iff)
-         (define-values (cond* cond-repr) (loop cond ctx))
-         (define-values (ift* ift-repr) (loop ift ctx))
-         (define-values (iff* iff-repr) (loop iff ctx))
-         (values `(if ,cond* ,ift* ,iff*) ift-repr)]
-        [`(! ,props ... ,body)
-         (define props* (props->dict props))
-         (loop body
-               (match (dict-ref props* ':precision #f)
-                 [#f ctx]
-                 [prec (struct-copy context ctx [repr (get-representation prec)])]))]
-        [`(cast ,body)
-         (define repr (context-repr ctx))
-         (define-values (body* repr*) (loop body ctx))
-         (if (equal? repr* repr) ; check if cast is redundant
-             (values body* repr)
-             (values (list (get-cast-impl repr* repr) body*) repr))]
-        [`(,(? constant-operator? x))
-         (define cnst (get-parametric-constant x (context-repr ctx)))
-         (values (list cnst) (impl-info cnst 'otype))]
-        [(list 'neg arg) ; non-standard but useful
-         (define-values (arg* atype) (loop arg ctx))
-         (define op* (get-parametric-operator 'neg atype))
-         (values (list op* arg*) (impl-info op* 'otype))]
-        [`(,op ,args ...)
-         (define-values (args* atypes) (for/lists (args* atypes) ([arg args]) (loop arg ctx)))
-         ;; Match guaranteed to succeed because we ran type-check first
-         (define op* (apply get-parametric-operator op atypes))
-         (values (cons op* args*) (impl-info op* 'otype))]
-        [(? variable?) (values expr (context-lookup ctx expr))]
-        [(? number?)
-         (define prec (representation-name (context-repr ctx)))
-         (define num
-           (match expr
-             [(or +inf.0 -inf.0 +nan.0) expr]
-             [(? exact?) expr]
-             [_ (inexact->exact expr)]))
-         (values (literal num prec) (context-repr ctx))])))
-  expr*)
+  (let loop ([expr (expand-expr prog)]
+             [prop-dict (repr->prop (context-repr ctx))])
+    (match expr
+      [(? number? n)
+       (literal (match n
+                  [(or +inf.0 -inf.0 +nan.0) expr]
+                  [(? exact?) expr]
+                  [_ (inexact->exact expr)])
```

Comment:
Neither here nor there but I don't know if this case ever occurs now. I think we now set up Racket to read decimals as exact, so I think they're all exact.

Reply:
(Or infinite. Anyway, I don't think this `match` ever does the last case, which is the case that matters.)